lib/node_modules/@stdlib/fs/resolve-parent-paths/lib/async.js

```diff
@@ -0,0 +1,372 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2024 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var resolve = require( 'path' ).resolve;
+var isStringArray = require( '@stdlib/assert/is-string-array' ).primitives;
+var isFunction = require( '@stdlib/assert/is-function' );
+var hasOwnProp = require( '@stdlib/assert/has-own-property' );
+var cwd = require( '@stdlib/process/cwd' );
+var exists = require( '@stdlib/fs/exists' );
+var format = require( '@stdlib/string/format' );
+var validate = require( './validate.js' );
+
+
+// VARIABLES //
+
+var MODES = {
+	'first': first,
+	'some': some,
+	'all': all,
+	'each': each
+};
+
+
+// FUNCTIONS //
+
+/**
+* Asynchronously resolves any first path from a set of paths by walking parent directories.
+*
+* @private
+* @param {Array<string>} paths - paths to resolve from
+* @param {string} dir - base directory
+* @param {Callback} done - callback to invoke after resolving paths
+* @returns {void}
+*/
+function first( paths, dir, done ) {
+	var child;
+	var spath;
+	var out = [];
+	var i;
+
+	// Start at a base directory and continue moving up through each parent directory...
+	for ( i = 0; i < paths.length - 1; i++ ) {
+		spath = resolve( dir, paths[ i ] );
+		if ( exists.sync( spath ) ) {
```

Comment:
@Snehil-Shah This is not async. This needs to be refactored to ensure that all operations are asynchronous. We have plenty of examples of async processing code from which to draw inspiration (e.g., see `utils/async/*` or `fs/read-file-list`). As is, this implementation has an odd mixture of sync and async logic. Only making the last check asynchronous and not all the others defeats the purpose of async operation.

In principle, we could allow concurrent checks, but that would complicate the logic a bit, especially during error conditions, such as too many files descriptors being open. We have some precedent for limiting concurrency and handling errors (e.g., `utils/async/*`), but I'm fine, for now, if we just support serially checking for existence.

Reply:
Yeah, I ended up making a mess trying to make sure one directory was fully traversed before moving on to the next. I changed it now..